### PR TITLE
Load metadata at start time not construction time

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we load the metadata from disk while constructing the node. However there
is no real need to do so, and this commit moves that code to run later while
the node is starting instead.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I've opened this against `7.x` not `master` because `7.x` has more interesting BWC issues, and a forward-port to remove them will be simpler than a backport to add them.

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 1]
```diff
@@ -71,41 +71,55 @@
...
  */
-public class GatewayMetaState implements ClusterStateApplier, CoordinationState.PersistedState {
+public class GatewayMetaState implements PersistedState {
     protected static final Logger logger = LogManager.getLogger(GatewayMetaState.class);
 
     private final MetaStateService metaStateService;
     private final Settings settings;
-    private final ClusterService clusterService;
-    private final TransportService transportService;
+    private final SetOnce<PersistedState> persistedState = new SetOnce<>();
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: I think this field requires a comment, explaining why we can not just return this (which implements PersistedState).

**Review by andrershov — APPROVED** [review]
I left one minor comment to add a comment, otherwise LGTM

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 2]
```diff
@@ -71,41 +71,55 @@
...
+        if (DiscoveryModule.DISCOVERY_TYPE_SETTING.get(settings).equals(DiscoveryModule.ZEN_DISCOVERY_TYPE)) {
+            // only for tests that simulate a mixed Zen1/Zen2 clusters, see Zen1IT
+            if (isMasterOrDataNode()) {
+                clusterService.addLowPriorityApplier(this::applyClusterState);
+            }
+            persistedState.set(new InMemoryPersistedState(getCurrentTerm(), getLastAcceptedState()));
+        } else {
+            if (DiscoveryNode.isMasterNode(settings) == false) {
+                if (DiscoveryNode.isDataNode(settings)) {
+                    // use Zen1 way of writing cluster state for non-master-eligible nodes
```
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: I would appreciate if you add an extended comment what is "Zen1" style of writing cluster state and is "Zen2" style of writing cluster.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Does this clarify? https://github.com/elastic/elasticsearch/pull/46326/files#diff-e5547fdd4d81fb437b07570d8c0de04fR118-R131

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: ++ although I hope we can remove the `implements PersistedState` in the near future. See https://github.com/elastic/elasticsearch/pull/46326/commits/59c50d0afa2f2a7771e8dcb4489f2da42fade6cf

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java (hunk 2 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: The explanation looks very good, thank you!

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java (hunk 1 above)
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Thanks, looks good

**Review by andrershov — APPROVED** [review]
LGTM

**Pull request merged**